MAINT Add one-sided set differences for clarity in param validation

Micky774 opened the pull request:
#### Reference Issues/PRs
Fixes #23744

#### What does this implement/fix? Explain your changes.
Add one-sided set differences for clarity in param validation

#### Any other comments?

Micky774 committed:
Added in one-sided set differences for clarity

Micky774 committed:
Improved wording a bit

glemaitre reviewed (approved):
LGTM

jeremiedbb commented:
Usually I run ``pytest -vl sklearn/tests/test_common.py -k check_param_validation`` to check that the constraints are properly set. It happens that this test also checks that the constraints match the parameters which will error before showing the message you added here. I guess it makes more sense to keep the message here so I'd remove the test for param match in ``check_param_validation``.

Micky774 committed:
Removed check for param match in test (now part of ValueError)

Micky774 committed:
Merge branch 'main' into param_val_sets

Micky774 committed:
Merge branch 'main' into param_val_sets

Micky774 commented:
@jeremiedbb Should be good now?

Micky774 committed:
Merge branch 'main' into param_val_sets

Micky774 committed:
Merge branch 'main' into param_val_sets

jeremiedbb commented:
I'm having second thoughts. I don't think that ``validate_params`` should ensure that there's a 1-to-1 match between the parameters and the dict of constraints. The reason is that it's a private thing and we don't want to enfore third party estimators to set or update ``_parameter_constraints``.

If a one writes a custom estimator that inherits from a sklearn estimator and calls ``super().fit``, it would error if the parameters are not the same. See this example:
```py
class MyEstimator(KMeans):
    def __init__(self, <all KMeans params>, additional_param):
        ...

    def fit(self, X, y=None):
        # do some extra stuff

        super().fit(X, y)  # <- calls self._validate_params and fails because 
                           # additional_param is not in _parameter_constraints
```

To not break user code (e.g. ImbalanceLearn maybe ? @glemaitre), I'd be in favor of being conservative and only validate params that are in the dict. We can still raise if a key in the dict doesn't correspond to any of the parameters.

For scikit-learn estimators we still want to make sure that there's a 1-to-1 match, which should be ensured by the common test.
What do you think @glemaitre @thomasjpfan ?

glemaitre commented:
> To not break user code (e.g. ImbalanceLearn maybe ? @glemaitre),

@jeremiedbb is right. I can think of some predictor like `BalancedRandomForestClassifier` with an additional parameter that will fail if we don't use the new tools.

It could be great to not impose it on third-party estimators for the moment.

Micky774 committed:
Merge branch 'main' into param_val_sets

Micky774 committed:
Updated tests/error to be more informative but not restrictive

Micky774 commented:
> To not break user code (e.g. ImbalanceLearn maybe ? @glemaitre), I'd be in favor of being conservative and only validate params that are in the dict. We can still raise if a key in the dict doesn't correspond to any of the parameters.

Updated this PR to implement this behavior. I added a list of unexpected/excess parameters to the error message as well to make it more informative. I moved the two set differences to the assertion error message in `check_param_validation`. Let me know what you think

thomasjpfan commented:
> For scikit-learn estimators we still want to make sure that there's a 1-to-1 match

I agree with not being strict for third party estimators and being strict for scikit-learn estimators.

jeremiedbb commented on sklearn/utils/_param_validation.py:
```diff
@@ -43,10 +43,11 @@ def validate_parameter_constraints(parameter_constraints, params, caller_name):
+            f" {set(parameter_constraints.keys())-set(params.keys())}"
         )
 
     for param_name, param_val in params.items():
```
We need to add something like
```py
        # We allow parameters to not have a constraint so that third party estimators
        # can inherit from sklearn estimators without having to necessarily use the
        # validation tools. 
        if param_name not in parameter_constraints:
            continue
```

otherwise it will fail if there's a parameter without constraint.

Please also add a test in `test_param_validation.py` for that case: a function with less constraints than params

Micky774 committed:
Addressed review

jeremiedbb commented on sklearn/utils/tests/test_param_validation.py:
```diff
@@ -403,11 +403,23 @@ def test_validate_params_match_error():
         pass
 
-    match = r"The parameter constraints .* do not match the parameters to validate"
+    match = r"The parameter constraints .* contain unexpected parameters {'c'}"
```
I think the docstring of this test can be updated to reflect the new behavior as well.

jeremiedbb commented on sklearn/utils/_param_validation.py:
```diff
@@ -43,13 +43,20 @@ def validate_parameter_constraints(parameter_constraints, params, caller_name):
         The name of the estimator or function or method that called this function.
     """
-    if params.keys() != parameter_constraints.keys():
+    if len(set(parameter_constraints.keys()) - set(params.keys())) != 0:
```
```suggestion
    if len(set(parameter_constraints) - set(params)) != 0:
```

jeremiedbb commented on sklearn/utils/_param_validation.py:
```diff
@@ -43,13 +43,20 @@ def validate_parameter_constraints(parameter_constraints, params, caller_name):
         raise ValueError(
-            f"The parameter constraints {list(parameter_constraints.keys())} do not "
-            f"match the parameters to validate {list(params.keys())}."
+            f"The parameter constraints {list(parameter_constraints.keys())}"
```
```suggestion
            f"The parameter constraints {list(parameter_constraints)}"
```

jeremiedbb commented on sklearn/utils/_param_validation.py:
```diff
@@ -43,13 +43,20 @@ def validate_parameter_constraints(parameter_constraints, params, caller_name):
-            f"match the parameters to validate {list(params.keys())}."
+            f"The parameter constraints {list(parameter_constraints.keys())}"
+            " contain unexpected parameters"
+            f" {set(parameter_constraints.keys())-set(params.keys())}"
```
linter is not complaining about spaces around operator ?
```suggestion
            f" {set(parameter_constraints) - set(params)}"
```

jeremiedbb reviewed (approved):
LGTM. Just a few remaining nitpcks

Micky774 committed:
Merge branch 'main' into param_val_sets

Micky774 committed:
Apply suggestions from code review

Co-authored-by: Jérémie du Boisberranger <34657725+jeremiedbb@users.noreply.github.com>

Micky774 committed:
Updated docstring

Micky774 committed:
Merge branch 'main' into param_val_sets

Micky774 commented:
@glemaitre Are you happy with the changes made after you first approved?

Micky774 committed:
Merge branch 'main' into param_val_sets

Micky774 commented:
@glemaitre @jeremiedbb Sorry for the spam, just wanted to follow up on this :)

Micky774 committed:
Merge branch 'main' into param_val_sets

ogrisel reviewed (approved):
LGTM as well.
